modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/trait/DistributionTrait.java

```diff
@@ -142,7 +142,7 @@
                     && DistributionFunction.satisfy(function, other.function));
 
         if (other.getType() == RANDOM_DISTRIBUTED)
-            return getType() == HASH_DISTRIBUTED;
+            return getType() == HASH_DISTRIBUTED || getType() == SINGLETON;
```

Comment:
I don't sure this is a legit change. If SINGLETON will satisfy a RANDOM distribution, then there will be no additional exchange, which leads to singleton be a part of fragment with RANDOM distribution, but this should not be possible, since this fragment is allowed to be executed on several nodes, whereas SINGLETON is expected to be executed on only one

Reply:
This change was made in the middle of the development, looks like it's not required anymore. I will revert it.

Reply:
I do not quite understand. If `single` means that the plan node is allowed to execute only on one cluster node, then, for the MAP phase any input with `single` distribution should be prohibited. Since, if all inputs are single - `IgniteSingleMinus` node should be used, and if there is a mix of single and random plan nodes - random plan nodes should be converted to single with `IgniteExchange`. So, `IgniteMapMinus.deriveDistribution` should convert all inputs to `random`, or skip this plan if there is at least one `single` node. 

I was thinking previously that `single` trait means that data for this plan node is located only on one cluster node (nevertheless plan node can be executed on different cluster nodes, the same for `broadcast`, it shows only data distribution, but not force plan node to be executed on every cluster node). In this case change with `single().satisfies(random())` looks fine. And this change makes possible some plans like:
```
IgniteReduceMinus(all=[false])
  IgniteExchange(distribution=[single])
    IgniteMapMinus(all=[false])
      IgniteTableScan(table=[[PUBLIC, SINGLE_TBL1]])
      IgniteTableScan(table=[[PUBLIC, RANDOM_TBL1]])
```
Should we prohibit such plans?

Reply:
> single trait means that data for this plan node is located only on one cluster node 

This is correct. And the plan above is correct too. But the problem caused by limitation of current splitter implementation. When an exchange connects two fragments with SINGLE and RANDOM distribution, and the data flows from the first to the second, the IgniteReceiver nodes becomes a source of the data with a SINGLE distribution, but presented on every node of the fragment with RANDOM distribution, thereby creates N copies of the original data.
The same problem with BROADCAST to RANDOM transition.

Reply:
I have an ideas hot cope with this problem, but for now we should not allows such plans.

Reply:
Ok, reverted this change for a while.